src/vtarget-out/inv-syn/grain_inv_parse.cc

```diff
@@ -0,0 +1,194 @@
+// if unsat --> add the (assert ...)
+bool GrainInvariantParser::ParseInvResultFromFile(const std::string & fname) {
+  std::ifstream fin(fname);
+  if (! fin.is_open()) {
+    ILA_ERROR << "Unable to read from : " << fname;
+    return false;
+  }
+/*
+  std::string result;
+  if (! std::getline(fin,result) ) {
+    ILA_ASSERT(false) << "No output from the synthesizer!" << result;
+    return false; // unknown result, possibly failed
+  }
+  do {
+    if (S_IN("unknown", result))
+      return false; // failed
+    if (S_IN("proved", result)) {
+      getline(fin,result); // the iter line, the rest are the invs
+      break;
+    }
+  } while (getline(fin,result));
+*/
+  std::stringstream sbuf;
+  sbuf << fin.rdbuf(); // different from original, it starts (define-fun)
+  raw_string = sbuf.str();
+
+  ParseSmtResultFromString(raw_string); // but the raw string are multiline and it should not be prune...
+  return true;
+}
+
+// -------------------------CALL BACK FUNS--------------------------------------------------------------------
+
+/// call back function to create a sort
+var_type * GrainInvariantParser::make_sort(const std::string &name, const std::vector<int> & idx) {
+  //ILA_ASSERT(! quantifier_def_stack.empty());
+  //ILA_ASSERT(! quantifier_var_def_idx_stack.empty());
+  
+  // should should only be BitVec or Bool
+  if (name == "Bool") {
+    if (! IN("Bool", sort_container))
+      sort_container.insert(std::make_pair(std::string("Bool"),
+        var_type(var_type::tp::Bool, 1, "")));
+    return & (sort_container["Bool"]);
+  }
+  else if (name == "BitVec") {
+    ILA_ASSERT(idx.size() == 1);
+    ILA_ASSERT(idx[0] > 0);
+    std::string sortIdxName = "BV"+std::to_string(idx[0]);
+    if (! IN(sortIdxName, sort_container))
+      sort_container.insert(std::make_pair(sortIdxName,
+        var_type(var_type::tp::BV, idx[0], "")));
+    return & (sort_container[sortIdxName]);
+  }
+  ILA_ASSERT(false) << "Unknown sort:" << name << " in flattened smt.";
+    return nullptr; // should not be reachable
+} // make_sort
+
+
+/// this is actually declare variables
+void GrainInvariantParser::declare_function(const std::string &name, var_type * sort) {
+  ILA_ASSERT(quantifier_def_stack.size() == 1) << "There should only be a global quantified var list";
+  // we need to extract the name from verilog
+  auto top_module = design_smt_info_ptr->get_module_def_orders().back();
+  /// we need to check the format
+  std::string vlg_name = name;
+  if (name.front() == '|' && name.back() == '|') {
+    ILA_ASSERT(name.size() > 2) << "Unexpected empty name: " << name;
+    vlg_name = name.substr(1,name.length()-2) ; // verilog-name should be extracted from the name part
+  }
+  if (vlg_name.find("S_") == 0) {
+    ILA_ASSERT(vlg_name.size() > 2) << "Unexpected empty name: " << name;
+    vlg_name = vlg_name.substr(2,name.length()-2);
+  }
+  std::string converted_name;
+  if (vlg_name.find(dut_verilog_instance_name+".") == 0) {  
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Inefficient usage of string::find() in condition; string::starts_with() could be faster.](https://app.codacy.com/app/Bo-Yuan-Huang/ILAng/pullRequest?prid=4555809)